File: shared/VisionFallbackTests.cs
using System;
using System.IO;
using System.Linq;
using DotsHarnessCore;
using HarnessPluginKit;
using Xunit;

namespace DotsHarness.Tests;

public sealed class VisionFallbackTests
{
    [Theory]
    [InlineData(400, "model does not support image input", true)]
    [InlineData(400, "model does not support images", true)]
    [InlineData(415, "multimodal input is not supported", true)]
    [InlineData(422, "vision is unsupported for this model", true)]
    [InlineData(400, "no image support on this endpoint", true)]
    [InlineData(401, "vision is not supported", false)]
    [InlineData(429, "image input is not supported", false)]
    [InlineData(500, "vision is not supported", false)]
    [InlineData(400, "invalid image payload", false)]
    [InlineData(400, "failed to decode image", false)]
    [InlineData(400, "vision model was not found", false)]
    [InlineData(404, "vision model is not available", false)]
    [InlineData(404, "unknown model", false)]
    [InlineData(400, "model does not support text input", false)]
    [InlineData(400, "bad request", false)]
    public void ClassifiesOnlyExplicitImageCapabilityFailures(int status, string message, bool expected)
    {
        Assert.Equal(expected, VisionProviderCapability.IsImageInputUnsupported(message, status));
    }

    [Fact]
    public void InstalledVisionPluginIsAddedToHostComposition()
    {
        var root = Path.Combine(Path.GetTempPath(), $"DotsHarnessVision-{Guid.NewGuid():N}");
        try
        {
            var paths = new PluginRuntime.SupportPaths(
                root,
                Path.Combine(root, "plugins"),
                Path.Combine(root, "presets"),
                Path.Combine(root, "settings.json"),
                Path.Combine(root, "host.patch.yml"),
                Path.Combine(root, "trust.json"),
                Path.Combine(root, "models"),
                Path.Combine(root, "runtime"));
            paths.Ensure();
            var folder = Path.Combine(paths.Plugins, VisionFallbackDefaults.PluginId);
            Directory.CreateDirectory(folder);
            File.WriteAllText(Path.Combine(folder, "plugin.yml"), """
                id: dots.vision-fallback
                name: Vision fallback
                version: 1.0.0
                abi: 1.0.0
                plane: host
                """);

            var catalog = new PluginRuntime.PluginCatalog(paths);
            catalog.Refresh();
            var document = CompositionLoader.LoadHost(paths, catalog);

            Assert.Contains(document.Entries, entry => entry.Plugin == VisionFallbackDefaults.PluginId);
        }
        finally
        {
            if (Directory.Exists(root)) Directory.Delete(root, recursive: true);
        }
    }
}

File: macos/DotsHarness/Tests/HarnessPluginKitTests/VisionFallbackTests.swift
import XCTest
import HarnessPluginKit
import PluginRuntime
import DotsHarnessCore

final class VisionFallbackTests: XCTestCase {
    func testClassifiesOnlyExplicitImageCapabilityFailures() {
        XCTAssertTrue(VisionProviderCapability.isImageInputUnsupported(
            message: "model does not support image input", statusCode: 400))
        XCTAssertTrue(VisionProviderCapability.isImageInputUnsupported(
            message: "model does not support images", statusCode: 400))
        XCTAssertTrue(VisionProviderCapability.isImageInputUnsupported(
            message: "multimodal input is not supported", statusCode: 415))
        XCTAssertTrue(VisionProviderCapability.isImageInputUnsupported(
            message: "no image support on this endpoint", statusCode: 400))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "vision is not supported", statusCode: 401))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "invalid image payload", statusCode: 400))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "failed to decode image", statusCode: 400))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "vision is not supported", statusCode: 500))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "vision model was not found", statusCode: 400))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "vision model is not available", statusCode: 404))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "unknown model", statusCode: 404))
        XCTAssertFalse(VisionProviderCapability.isImageInputUnsupported(
            message: "model does not support text input", statusCode: 400))
    }

    @MainActor
    func testInstalledVisionPluginIsAddedToHostComposition() throws {
        let paths = temporaryPaths()
        let folder = paths.plugins.appendingPathComponent(VisionFallbackDefaults.pluginID, isDirectory: true)
        try FileManager.default.createDirectory(at: folder, withIntermediateDirectories: true)
        try """
        id: dots.vision-fallback
        name: Vision fallback
        version: 1.0.0
        abi: 1.0.0
        plane: host
        """.write(to: folder.appendingPathComponent("plugin.yml"), atomically: true, encoding: .utf8)

        let catalog = PluginCatalog(paths: paths)
        catalog.refresh()
        let document = CompositionLoader.loadHost(paths: paths, catalog: catalog)

        XCTAssertTrue(document.entries.contains { $0.plugin == VisionFallbackDefaults.pluginID })
    }

    @MainActor
    private func temporaryPaths() -> SupportPaths {
        let root = FileManager.default.temporaryDirectory
            .appendingPathComponent("DotsHarnessVisionTests-\(UUID().uuidString)", isDirectory: true)
        let paths = SupportPaths(
            root: root,
            plugins: root.appendingPathComponent("plugins", isDirectory: true),
            presets: root.appendingPathComponent("presets", isDirectory: true),
            settings: root.appendingPathComponent("settings.json"),
            hostPatch: root.appendingPathComponent("host.patch.yml"),
            trust: root.appendingPathComponent("trust.json"),
            models: root.appendingPathComponent("models", isDirectory: true),
            runtime: root.appendingPathComponent("runtime", isDirectory: true)
        )
        paths.ensure()
        addTeardownBlock { try? FileManager.default.removeItem(at: root) }
        return paths
    }
}
